File: TcpSocketSystem_Java/src/com/tecnm/sockets/usecases/gestionarsesiones/GestionarSesionesHandler.java
package com.tecnm.sockets.usecases.gestionarsesiones;

import com.tecnm.sockets.domain.modelos.EstadoSesion;
import com.tecnm.sockets.domain.puertos.IRepositorioSesiones;
import com.tecnm.sockets.domain.puertos.IServicioLogging;

/**
 * Implementación de la gestión y auditoría de sesiones concurrentes en Java.
 */
public final class GestionarSesionesHandler implements IGestionarSesionesUseCase {

    private final IRepositorioSesiones repositorioSesiones;
    private final IServicioLogging logger;

    public GestionarSesionesHandler(IRepositorioSesiones repositorioSesiones, IServicioLogging logger) {
        if (repositorioSesiones == null || logger == null) {
            throw new IllegalArgumentException("Las dependencias no pueden ser nulas.");
        }
        this.repositorioSesiones = repositorioSesiones;
        this.logger = logger;
    }

    @Override
    public EstadoSesion registrarNuevaConexion(String sesionId, String direccionIp, int puerto) {
        EstadoSesion sesion = new EstadoSesion(sesionId, direccionIp, puerto);
        repositorioSesiones.registrarSesion(sesion);

        logger.auditoria("CLIENTE_CONECTADO", direccionIp + ":" + puerto, "Sesión ID: " + sesionId + " iniciada.");
        logger.info(String.format("[CONEXIÓN ABIERTA] Cliente %s:%d (ID: %s). Activos: %d",
                direccionIp, puerto, sesionId, repositorioSesiones.obtenerTotalSesionesActivas()), "GestionarSesionesHandler");

        return sesion;
    }

    @Override
    public void registrarActividadMensaje(String sesionId) {
        EstadoSesion sesion = repositorioSesiones.obtenerPorId(sesionId);
        if (sesion != null && sesion.estaActiva()) {
            sesion.incrementarMensajes();
            repositorioSesiones.actualizarSesion(sesion);
        }
    }

    @Override
    public void registrarDesconexion(String sesionId, String motivo) {
        EstadoSesion sesion = repositorioSesiones.obtenerPorId(sesionId);
        if (sesion != null) {
            sesion.cerrarSesion();
            repositorioSesiones.finalizarSesion(sesionId);

            String motivoFinal = (motivo != null) ? motivo : "Desconexión normal";
            logger.auditoria("CLIENTE_DESCONECTADO", sesion.getDireccionRemota() + ":" + sesion.getPuertoRemoto(),
                    "Sesión " + sesionId + " finalizada. Motivo: " + motivoFinal + ". Mensajes: " + sesion.getMensajesProcesados());
            logger.info(String.format("[CONEXIÓN CERRADA] Cliente %s:%d (ID: %s). Activos: %d",
                    sesion.getDireccionRemota(), sesion.getPuertoRemoto(), sesionId, repositorioSesiones.obtenerTotalSesionesActivas()), "GestionarSesionesHandler");
        }
    }

    @Override
    public int obtenerConexionesActivas() {
        return repositorioSesiones.obtenerTotalSesionesActivas();
    }
}

File: TcpSocketSystem_DotNet/src/UseCases/GestionarSesiones/GestionarSesionesHandler.cs
using TcpSocketSystem.Core.Domain.Modelos;
using TcpSocketSystem.Core.Domain.Puertos;

namespace TcpSocketSystem.Core.UseCases.GestionarSesiones;

/// <summary>
/// Implementación del caso de uso de gestión de sesiones concurrentes con auditoría.
/// </summary>
public sealed class GestionarSesionesHandler : IGestionarSesionesUseCase
{
    private readonly IRepositorioSesiones _repositorioSesiones;
    private readonly IServicioLogging _logger;

    public GestionarSesionesHandler(IRepositorioSesiones repositorioSesiones, IServicioLogging logger)
    {
        _repositorioSesiones = repositorioSesiones ?? throw new ArgumentNullException(nameof(repositorioSesiones));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public EstadoSesion RegistrarNuevaConexion(string sesionId, string direccionIp, int puerto)
    {
        var sesion = new EstadoSesion(sesionId, direccionIp, puerto);
        _repositorioSesiones.RegistrarSesion(sesion);

        _logger.Auditoria("CLIENTE_CONECTADO", $"{direccionIp}:{puerto}", $"Sesión ID: {sesionId} iniciada exitosamente.");
        _logger.Info($"[CONEXIÓN ABIERTA] Cliente {direccionIp}:{puerto} (ID: {sesionId}). Clientes activos: {_repositorioSesiones.ObtenerTotalSesionesActivas()}", nameof(GestionarSesionesHandler));

        return sesion;
    }

    public void RegistrarActividadMensaje(string sesionId)
    {
        var sesion = _repositorioSesiones.ObtenerPorId(sesionId);
        if (sesion != null && sesion.EstaActiva)
        {
            sesion.IncrementarMensajes();
            _repositorioSesiones.ActualizarSesion(sesion);
        }
    }

    public void RegistrarDesconexion(string sesionId, string motivo = "Desconexión normal")
    {
        var sesion = _repositorioSesiones.ObtenerPorId(sesionId);
        if (sesion != null)
        {
            sesion.CerrarSesion();
            _repositorioSesiones.FinalizarSesion(sesionId);

            _logger.Auditoria("CLIENTE_DESCONECTADO", $"{sesion.DireccionRemota}:{sesion.PuertoRemoto}", $"Sesión {sesionId} cerrada. Motivo: {motivo}. Mensajes procesados: {sesion.MensajesProcesados}");
            _logger.Info($"[CONEXIÓN CERRADA] Cliente {sesion.DireccionRemota}:{sesion.PuertoRemoto} (ID: {sesionId}). Clientes activos: {_repositorioSesiones.ObtenerTotalSesionesActivas()}", nameof(GestionarSesionesHandler));
        }
    }

    public int ObtenerConexionesActivas()
    {
        return _repositorioSesiones.ObtenerTotalSesionesActivas();
    }
}
